### ISSUE TITLE
Subscribe: Atomically collect entire config in a single python object?

### ISSUE THREAD

**Issue by esend7881** [sysrepo/sysrepo#2139]
## Issue

Inside a module subscription function, how can I extract the full config as a single object in a subscribe loop after an `edit-config` atomically?

## Details

I am using the standard python `module_change_cb` function to collect changing config data:

```
def module_change_cb(sess, module_name, event, private_ctx):
    try:
        if (sr.SR_EV_APPLY == event):
            change_path = "/" + module_name + ":*"
            it = sess.get_changes_iter(change_path);
            while True:
                change = sess.get_change_next(it)
                if change == None:
                    break
                print_change(change.oper(), change.old_val(), change.new_val())

   
    except Exception as e:
        print e

    return sr.SR_ERR_OK
```

What I am noticing is each individual `leaf` of the config is reported in the `while True` loop. Is there a way to have the `change.new_val()` and `change.old_val()` be a single json-type object that contains the entire configuration?

I understand under normal usage one would want to know the individual leafs and elements that have changed, but it would actually be simpler if the entire config is given to me whole.

## If not possible

If this isn't possible or violates a best-practice of some kind, do you recommend a way to collect all the new-config data so I can manually create my own single object to hold it all in? The way it is programmed now, I wouldn't know how to separate two-or-more `edit-config` pushes to the same config. 

## Environment:

- sysrepo 1.4.70
  - compiled with `cmake -DGEN_LANGUAGE_BINDINGS=ON ..`
- netopeer2 1.1.39 
  - compiled with `cmake -DSYSREPO_TIMEOUT=119 -DRPC_TIMEOUT=120 ..`
- libyang 1.0.184 
  - compiled with `cmake -DGEN_LANGUAGE_BINDINGS=ON ..`
- CentOS 7
- python 2.7.5/3.6.8

**Comment by smalls12** [comment]
Wouldn't you be able to do just a regular get_items?

**Comment by smalls12** [comment]
Ah I see you still want to know old and new values.
Nevermind then :)

**Comment by esend7881** [comment]
@smalls12 - will the `get_items` trigger when a new config is issued via `edit-config` though?

**Comment by smalls12** [comment]
The module change callback itself indicates that something has changed.

The get_items or get_changes are just mechanisms to view the data.

get_changes will only show the explicit changes.

get_items just shows what's currently there and won't indicate where the change was.

**Comment by michalvasko** [comment]
Hi,
if you use `get_data()`, you will get one data tree with the selected **new** data. If you create a new session in the callback and do the same, you will actually get the **old** data, so that is what you can do.

**Comment by esend7881** [comment]
@michalvasko - thanks for the tip regarding `get_data`. I am looking at the python object that is returned by `get_data` and see that much of the data is hidden by SWIG.

I believe the C object backing the get_data is from this class:
https://github.com/CESNET/libyang/blob/v1.0.184/swig/cpp/src/Tree_Data.hpp#L103

And I'm struggling to find how to extract the actual data out of the object. The closest I can come up with is this:
```
        for i in range(sess.get_items(change_path).val_cnt()):
            logging.info("val %s: %s", i, sess.get_items(change_path).val(i).to_string())
```
This is just junky debugging code.

Also, I notice that I need to send in a `replace` for my edit-config command in order to only see the newest data.

That all said, I do feel like there has got to be a more elegant way to extra the data via SWIG. If not, I suppose I can extract the data via a regex from the `to_string()` method

**Comment by michalvasko** [comment]
Hi,
I think you are better off using the [CFFI Python bindings](https://github.com/sysrepo/sysrepo-python) if you do not like the SWIG ones (which is understandable). Also, it seems to me you want the full final edit-config (which is called `diff` in *sysrepo*). That is exactly what is being read by `get_change_next` but it is not available directly in *sysrepo* API because I did not think there is any use-case for it.

**Comment by esend7881** [comment]
@michalvasko - thank you. Yes I am finding the SWIG based API is not giving me data that is naturally parse-able to a full JSON or XML object. I've actually spent the last day and a half just trying to merge all of it together and it is a mess!

I am very intrigued by these CFFI Python Bindings and will try it now. Will v0.3.1 (https://github.com/sysrepo/sysrepo-python/releases/tag/v0.3.1) properly correlate with my versions:

- sysrepo 1.4.70
  - compiled with cmake -DGEN_LANGUAGE_BINDINGS=ON ..
- netopeer2 1.1.39
  - compiled with cmake -DSYSREPO_TIMEOUT=119 -DRPC_TIMEOUT=120 ..
- libyang 1.0.184
  - compiled with cmake -DGEN_LANGUAGE_BINDINGS=ON ..
- CentOS 7
- python 2.7.5/3.6.8

In other words, do you use those same versions of sysrepo and libyang and use the same CMAKE parameters?

**Comment by michalvasko** [comment]
Hi,
> But that said, is there a way (perhaps through libyang or sysrepo cmake arguments) to have the full edit-config returned?

I have already told you that this is not possible this way. But, you can subscribe to the `edit-config` RPC and then you will always get the whole RPC content with its edit content.

**Comment by esend7881** [comment]
@michalvasko - I will look into RPC callbacks for edit-config today.

By the way, I noticed that the callback for `subscribe_module_change` doesn't return any values. I.e., with the SWIG based functions, I could return `sr.SR_EV_ABORT` or `sr.SR_ERR_OK` to tell sysrepo whether or not to accept the config. However, when I look at https://github.com/sysrepo/sysrepo-python/blob/master/examples/application.py#L66 there is no return value or obvious way to tell sysrepo to accept/reject the config.

**Comment by michalvasko** [comment]
You should ask these question in the `sysrepo-python` project since it is a stand-alone project maintained by another developer.

**Comment by esend7881** [comment]
Okay I think I found the full XML using the old style SWIG api!

```
subscribe.rpc_subscribe("/ietf-netconf:edit-config", rpc_cb, None)
...
def rpc_cb(sess, xpath, vals, event_id, request_id, vals_holder, private_data):
    pydevd.settrace(suspend=False, trace_only_current_thread=True)
    xml = vals.val(5).val_to_string()
    xml_object = vkb.xml(xml)
    logging.debug("XML: %s", xml_object)
    return sr.SR_ERR_OK
```

What I notice is that `vals.val(5)` contains the full XML of the config submitted through `edit-config`. It'd be nice if there was a more direct way to extract it, as I randomly found it as element #5 when I parsed through the data in the callback, but it's there!

Also, returning `SR_ERR_OK` or `SR_EV_ABORT` does the trick in telling sysrepo whether or not to accept/reject.

@michalvasko - thank you for the hints.

**Comment by michalvasko** [comment]
That the 5th value essentially gives you the full XML is just a coincidence so I do not think it is a solution, but whatever. Also, please return only `SR_ERR_*` values from callbacks, `SR_EV_ABORT` is an event type, not an error value.

**Comment by esend7881** [comment]
Ah, I assume I need to use one of these values:
https://github.com/sysrepo/sysrepo/blob/7920e6a336cf9d266cac17d5b459f0d9d42af754/src/sysrepo.h#L51

Thanks again for your help


As for element #5 xml, I ended up creating a regex search for namespace and
xml validity as I didn't want to assume it would always be element #5.

>
